### PR TITLE
[local gc] gc spinning part 1

### PR DESCRIPTION
**Pull request by davmason**
There isn't enough time left before ZBB to finish tuning the GC spinning. This PR adds the APIs so that the interface has the methods and we are free to switch to using the normalized spin methods whenever we have a chance to do the work

### CODE REVIEW

File: src/vm/gcenv.ee.cpp [hunk 1]
```diff
@@ -397,6 +397,16 @@ Thread* GCToEEInterface::GetThread()
     return ::GetThread();
 }
 
+void GCToEEInterface::YieldProcessorNormalized()
+{
+    ::YieldProcessorNormalized(YieldProcessorNormalizationInfo());
+}
+
+void GCToEEInterface::EnsureYieldProcessorNormalizedInitialized()
```
Comment (jkotas) [inline src/vm/gcenv.ee.cpp]: How is this going to be used? The extra condition to verify that the counts were initialized is pretty small relative to the call overhead. I think you can just check for the initialization inside YieldProcessorNormalized. No need to have a special API for it.

**Comment by Maoni0** [comment]
we should have an API that just returns the number that the VM already calculated though...whenever the VM is done calculating it can call an API to inform GC that number. please see https://github.com/dotnet/coreclr/issues/13983#issuecomment-376705087.

**Comment by davmason** [comment]
@Maoni0 I saw the comment, but it didn't seem necessary to go through the extra effort. It seems like we can absorb the extra overhead of the call in whatever tuning we do. According to Kount's numbers the pause delay is up to 140 cycles in post skylake, and a virtual call should be a small fraction of that. Do you anticipate the overhead of the call causing problems?

File: src/vm/gcenv.ee.cpp (hunk 1 above)
Comment (kouvel) [reply src/vm/gcenv.ee.cpp]: Even if you want to use an interface call per YieldProcessorNormalized(), this shouldn't be necessary. The first few calls (before initialization) will use default values and subsequent calls with use initialized values.

**Comment by kouvel** [comment]
Is the GC to EE interface shared with the desktop framework? The normalization done in desktop is different from core so that may complicate things a bit. Making an interface call per YieldProcessorNormalized() could keep it generic between the two with a slight overhead but the tuning between the runtimes may need to be different anyway.

**Comment by kouvel** [comment]
Although, the signature for YieldProcessorNormalized() is also different between core and desktop. Maybe it would be ok to use different strategy for each.

**Comment by Maoni0** [comment]
I am not sure we can say it's a small fraction - pre Skylake YieldProcessor takes a lot less and with CFG a virtual call can be pretty expensive. regardless, I don't see any reason why we would need to make a call at all...it's just a number we need to get from the VM side right? @kouvel I have not looked at desktop implementation so do you mind giving a brief explanation what the difference is (does it calculate a number that could change)?

**Comment by kouvel** [comment]
The normalization changes that went into desktop are different:
- CoreClr normalizes the delay to something close to that on Skylake processors. The increase in delay on pre-Skylake processors was compensated for by retuning spin loops.
- Desktop normalizes the delay to something close to that on pre-Skylake processors. This doesn't need to retune spin loops but a single yield would still take longer on Skylake processors. So the signature there is YieldProcessorNomralized(unsigned int count) and the total delay is adjusted based on one-time measurement.

> I don't see any reason why we would need to make a call at all...it's just a number we need to get from the VM side right?

In core, there is g_yieldsPerNormalizedYield, which is initialized to a default value and updated after measurement at the start of the finalizer thread. So if you want to store the value and not make interface call anymore, it would need to sync with the initialization by blocking or polling.

In desktop, there is a different measurement g_measured100PsPerYield. The count passed in to YieldProcessorNormalized() is scaled based on that value and a constant. Initialization works similarly to core.

The interface call could be avoided most of the time with something like this:

```c++
void YieldProcessorNormalized(unsigned int count)
{
    unsigned int normalizationInfo;
    if (g_isYieldProcessorNormalizedInitialized)
    {
        normalizationInfo = g_yieldProcessorNormalizationInfo;
    }
    else
    {
        g_yieldProcessorNormalizationInfo = normalizationInfo = interface->GetYieldProcessorNormalizationInfo(&g_isYieldProcessorNormalizedInitialized);
    }

    // code cloned from core and desktop, code is different for each
}
```

If it's ok to block for initialization, GetYieldProcessorNormalizationInfo() wouldn't need the first parameter but it would need to be called in some place before any yield activities and on a background thread so that startup time is not affected.

Or if an interface call per set of yield iterations is ok, then could do something like this:

```c++
interface->YieldProcessorNormalized(count);
```

In both cases the interface call could be shared between core and desktop.

**Comment by Maoni0** [comment]
thanks. 

if the initialization works similar between desktop and core, doesn't that just mean we can have a variable on the GC side that's set to some initial value; and the VM side can just make a call to the GC side to update that var whenever it finishes calculating it? GC can use the initial value up till then which is not a big deal 'cause it's such a short amount of time.

I am also not sure why there's such a difference on desktop and on core...can you please elaborate the rationale behind it?

**Comment by kouvel** [comment]
> doesn't that just mean we can have a variable on the GC side that's set to some initial value; and the VM side can just make a call to the GC side to update that var whenever it finishes calculating it?

That's even better

> I am also not sure why there's such a difference on desktop and on core...can you please elaborate the rationale behind it?

Several changes went into core related to this, some of them quite large (esp. monitor changes). Spinning behavior through Thread.SpinWait() would also change and may necessitate users to retune their spin loops as well. The issue mainly affects spin loops that are very long or frequent. We wanted to limit the risk of regression in desktop on older processors. The changes in core are better, the longer delay actually allows for some reasonable work to be done such that it's more likely for a change to be observed after one yield.

**Comment by Maoni0** [comment]
thanks for the explanation, @kouvel!

**Comment by davmason** [comment]
@Maoni0 @kouvel I changed the api so that the VM initializes it. The default value of g_yieldProcessorNormalizationInfo is 1 which seems like a safe default

File: src/gc/gcinterface.h [hunk 2]
```diff
@@ -720,6 +720,9 @@ class IGCHeap {
     // Tells the GC when the VM is suspending threads.
     virtual void SetSuspensionPending(bool fSuspensionPending) = 0;
 
+    // Tells the GC how many YieldProcessor calls are equal to one normalized yield processor call.
+    virtual void SetYieldProcessorNormalizationInfo(uint32_t yieldProcessorNormalizationInfo) = 0;
```
Comment (jkotas) [inline src/gc/gcinterface.h]: This `YieldProcessorNormalizationInfo` does not match what we call YieldProcessorNormalizationInfo in the VM (it is 3 numbers in the VM), and it is does not really tell what this is.

`YieldProcessorScalingFactor` would be a better name for this, I think.

Comment (Maoni0) [reply src/gc/gcinterface.h]: SetYieldProcessorScalingFactor sounds good to me

**Pull request merged**